test/evp_test.c

```diff
@@ -2524,11 +2597,7 @@ static int pkey_test_run(EVP_TEST *t)
     OPENSSL_free(got);
     got = NULL;
 
-    /* Repeat the test on a copy. */
-    if (!TEST_ptr(copy = EVP_PKEY_CTX_dup(expected->ctx))) {
-        t->err = "INTERNAL_ERROR";
-        goto err;
-    }
+    /* Repeat the test on the EVP_PKEY context copy. */
```

Comment:
Can you explain why this late dup worked before and not now?

Reply:
Yes.

It worked previously because this was only used to test `EVP_PKEY_sign()` that was initialized with `EVP_PKEY_sign_init_ex()`, i.e. where the MD is never actually used to digest anything, it's only there to check the *tbslen* against the MD output size, and to get the correct OID into the PKCS#1 formatted message to be signed.

Now that this function is also used to test `EVP_PKEY_sign()` that was initialized with `EVP_PKEY_sign_message_init()`, the the MD is actually used to digest the message, and its state is modified before this duplication, causing the repeat test to fail (obviously).

Reply:
And look, who knows what state changes will happen with the first expected->keyop()?  Only the providers will really know.  So, duplicating this late, and expecting the same result in the repeat test is just bad form.  I consider it pure luck that it worked 'til now.

Reply:
This morning, I realised that what I wrote above isn't quite right.  From the docs NOTES section in the docs, `EVP_PKEY_sign()` and `EVP_PKEY_sign_message_final()` should reset the context so another signature can be performed.  By extension, a late dup like this should be possible, because it should duplicate a context that's already reset to its original state.

So it turns out that your question leads to finding a but in the RSA implementation.

Reply:
On the other hand, perhaps it's wrong to pass on the pre-hashed message semantics to the message semantics.

From a practical point of view in a provider, it's going to be easier not to have the same reset semantics in the message calls, because you have to consider what parameters were passed in the init function, even more so that we also pass them to the MD init function, which has to be done again every time a reset needs to be performed.

So, I'm leaning toward modifying the documentation so that the reset semantics isn't guaranteed for the (not pre-hashed) message functionality.